New extensible logging features

sirreal opened the pull request:
Please continue any discussion on #12716.

Implements #10468. This PR replaces #12340.

Todo:
- [x] Find older PHPUnit compatible replacement for `withConsecutive` (PHP 5.2 travis tests fail)
- [x] Return single WC_Logger instance from `wc_get_logger`
- [x] Add bulk actions (delete) to db log list view
- [x] [Clarify ~`wc_safe_dump`~: improve name and description](https://github.com/woocommerce/woocommerce/pull/12340#pullrequestreview-11567587) now `wc_print_r`
- [x] [Throttle messages from email handler](https://github.com/woocommerce/woocommerce/pull/12340#pullrequestreview-11819674). Add a message about limitations of handler and aggregate/batch send messages once per request.

sirreal committed:
Extract file log handler

sirreal committed:
WC_Logger registers and delegates logging to handlers

sirreal committed:
Register file log handler by default

Add wc_register_file_log_handler function.
Calls add_filter with register function to include handler by default

sirreal committed:
Update tests for WC_Logger

sirreal committed:
Add abstract WC_Log_Handler class

Abstract base class is implemented with default formatting implementations.
`$timestamp` has been added to the `WC_Log_Handler::handle` implementation. This ensures all handlers are passed the same timestamp for the same message.
`_legacy` context has been added to `WC_Log_Handler_File` to ensure handle deprecated `WC_Logger::add` provides the same log entries as before.

Fix CS problems.

sirreal committed:
Add logic for log level filtering to `WC_Log_Handler`.

sirreal committed:
Adds tests for bubbling, etc.

Add tests teardown to cleanup log files
Fix errors in log handlers

sirreal committed:
Remove 2nd filter parameter that would never be called

sirreal committed:
Add tag context test for file log handler

sirreal committed:
Use wp_parse_args for default args

sirreal committed:
Cleanup "A" test log file

sirreal committed:
Remove redundant __constructor

sirreal committed:
Add email log handler

sirreal committed:
Extract anonymous functions to improve test compatibility.

sirreal committed:
Provide basic messages for email log handler

sirreal committed:
Fix tests for compatibility with PHP5

sirreal committed:
Restore deprecated functions WC_Logger::clear & WC_Logger::remove

Also updates clear test to supply required argument.

sirreal committed:
Use older test mock syntax for legacy compatibility.

sirreal committed:
Fix docblocks

sirreal committed:
Fix scrutinizer issues

sirreal committed:
Change defualt log handler threshold to DEBUG following discussion with team members.

sirreal committed:
Fix file handle regression

sirreal committed:
Add `setMethods` to test mocks

This may improve php 5.2 test failures

sirreal committed:
Add test for log() === [level_short_method]()

sirreal committed:
Defauly log file level is DEBUG

sirreal committed:
Restructure log tests to own directory

Also moves parent::tearDown to end of tearDown
Fix phpcs problems.

sirreal committed:
Fix tests for PHP 5.2

sirreal committed:
Remove _ from private/protected properties

sirreal committed:
Remove optional from required log handler context array

sirreal committed:
Improve log email message format.

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#pullrequestreview-8895028

sirreal committed:
Include site name in log email message

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#discussion_r88213040

sirreal committed:
Simplify default log entry format

Remove punctuation. Space separate values:

TIMESTAMP LEVEL MESSAGE

sirreal committed:
Improve logic for legacy `->add` message formatting

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#discussion_r88318220

sirreal committed:
Restore log deletion from admin log viewer page

WC_Logger no longer controls logging and has no knowledge of files.
Correctly handle `remove_log` action via file log handler.

sirreal committed:
Fix phpcs warning

sirreal committed:
Prevent possible undefined index warning

sirreal committed:
Add specific file log handler tests

sirreal committed:
Rotate logs

WC_Log_Handler_File now includes log rotation.
Add and improve tests.

sirreal committed:
Improve handle signature

The signature of WC_Log_Handler::handle mirrors log messages which is more intuitive.
Method docblocks have also been improved and are more complete.

sirreal committed:
Prefer PHP_EOL over "\n"

sirreal committed:
Add DB log handler

sirreal committed:
Update @package for log handlers

sirreal committed:
Implements database log viewer

sirreal committed:
Log viewer via html template

sirreal committed:
Add level views to db log table

sirreal committed:
Version 1.0.0 for abstract wc log handler

sirreal committed:
Provide correct class name to woocommerce_log_handler_set_threshold filter

sirreal committed:
Add woocommerce_format_log_entry filter to base WC_Log_Handler class

sirreal committed:
Add wc_safe_dump function

Some "safe" server configurations disable certain functions. `wc_safe_dump` attempts to
safely call a series of alternatives suitable for dumping an expression.

sirreal committed:
Fix phpcs issues

sirreal committed:
Update paypal gateway log method

sirreal committed:
update paypal gateway logs

sirreal committed:
remove print_r calls from paypal handler

sirreal committed:
Update logger calls

sirreal committed:
Update print_r to use wc_safe_dump

sirreal committed:
Include missing 5 in log level severities

sirreal committed:
Add severity level tests

sirreal committed:
Add log flushing to db log viewer

sirreal committed:
Mark db handler add method as static

sirreal committed:
Add DB handler tests

sirreal committed:
Add confirmation to db log flush

sirreal committed:
Remove bad logger test

Test comparing equality of logger results could randomly fail depending on execution time.
The coverage was duplicated by other tests, test was redundant and has been removed.

sirreal committed:
Fix misnamed test class

sirreal committed:
Rename test file log -> logger

sirreal committed:
Prefer "unit-tests" tag for testing log handler

sirreal committed:
Add test for multiple file log handlers writing to same file.

sirreal committed:
Add basic email logger test

sirreal committed:
Add woocommerce_default_log_handler setting

Adds configuration to settings page
Show status-logs page based on settings (unify logs & database logs)
Register default log handler by default

sirreal committed:
Merge pull request #12511 from sirreal/wc-logger

Merge WC_Logger implementation proposal

sirreal committed:
Fix logs-db redirects

sirreal committed:
Use autoloader for log handlers

sirreal committed:
Explicit use of file log handler for logger tests

sirreal committed:
Fix spelling errors

sirreal committed:
Merge pull request #12512 from sirreal/wc-logger

Wc logger

sirreal committed:
Spelling and doc fixes

sirreal committed:
Drop `var_dump` as safe dump alternative

var_dump does not support returning string representation.

sirreal committed:
Move log levels to independent class

sirreal committed:
Drop "bubbling" support for logs

The concept of bubbling or consuming logs has little benefit for our use case.
Drop support for bubbling and simplify logging.

sirreal committed:
Add more information about log levels

sirreal committed:
Remove threshold logic from handlers.

Email handler implements threshold logic internally. Other handlers handle all logs received.
Handler constructors have changed.

sirreal committed:
Add log threshold setting to settings page

sirreal committed:
Accept explicit handlers and threshold in WC_Logger

sirreal committed:
Improve logger tests

sirreal committed:
Improve logger tests

sirreal committed:
Drop WP_DEBUG override threshold. Improve logger tests.

This may be unexpected behavior for a user, their level is overridden with no
way to change it. It also greatly complicates testing default thresholds.

Dependency on WC_Log_Handler_File has been removed from logger tests.

sirreal committed:
Fix spelling

sirreal committed:
Include WC_Log_Handler with other abstracts

sirreal committed:
Store log levels as integer representations in database

This will allow indexed ordering by severity.
Add WC_Log_Levels::get_severity_level method to translate severity
integers to level strings. The inverse of WC_Log_Levels::get_level_severity.
Add context to tests.

sirreal committed:
Add email log handler tests

sirreal committed:
Make lot threshold strings translatable

sirreal committed:
Merge pull request #12574 from sirreal/wc-logger

Pull logger changes

dwainm commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,135 @@
+			return wp_mail( $this->recipients, $subject, $body );
+		}
+
+		return true;
```
`true` will be the default, even if it wasn't handled?

sirreal replied:
The return signature here is `@return bool True on success`, so `true` and `false` represent success and failure, respectively. The `true` in this case says essentially represents the success of no-op. We decided not to handle the log, and were successful in that. This is confusing if you interpret the return signature as "I did something with the log".

From a practical standpoint, the return value is [currently ignored by the logger](https://github.com/woocommerce/woocommerce/blob/54d131da39b00469ca104dcff828e830546aae25/includes/class-wc-logger.php#L110-L112). In the case that the logger decides to use the value in the future, it's hard to anticipate whether it's better to indicate "success/failure" or "handled/not handled (maybe failed?)". It's also hard to represent these nuances with the binary `true`/`false`.

It may be best to remove the return value entirely and add it when and if we have a clear use case in the future. What are your thoughts?

dwainm replied:
When I think of it in terms of how WP does [update_option](https://codex.wordpress.org/Function_Reference/update_option), `true` means something was saved/change and false means nothing happened. 

I'm not sure if we'll ever need it, but it should mean something very specific like: 
`true`  - the logger handled the request successfully 
`false` - the logger did not handle the request [succesfully]

Just some thoughts.

sirreal replied:
The perspective and in particular the comparison are helpful. I'll change this to `false` and update the return description to be similar to `update_option`. That seems most obvious and matches existing patterns.

Commit:
Improve @return signature on WC_Log_Handler::handle

Make meaning of returned bool explicit.
Update WC_Log_Handler_Email::handle accordingly.

via https://github.com/woocommerce/woocommerce/pull/12575#pullrequestreview-12381573

sirreal committed:
Return single instance from wc_get_logger

Remove class include from wc_get_logger (should be autoloaded).
Add tests for wc_get_logger.

sirreal committed:
Add delete to db log viewer table

Clean up some nonce logic.
Use a single <form> containing log table.

sirreal committed:
Mark WC_Log_Handler::format_entry abstract

sirreal committed:
Aggregate email handler messages (max: 1 email/request/handler)

Update emails accordingly.
Update email handler tests.

sirreal committed:
Protect methods

sirreal committed:
Improve test compatibility (PHP 5.2)

sirreal committed:
Improve PHP 5.2 test compatibility (older PHPUnit version)

sirreal committed:
Change wc_safe_dump to wc_print_r

Match print_r arguments, this makes usage obvious.

sirreal committed:
Add unit tests for wc_print_r()

Clean up some log logic ( `assertEquals( true ... )` becomse `assertTrue( ... )` ).

sirreal committed:
Fix WC_Log_Levels::is_valid_level tests

The wrong method was being called in tests.

sirreal committed:
Cleanup log list table view

Format log timestamp.
Add is_valid_level check to $_REQUEST level filter.
General cleanup: formatting, layout, commentary.

sirreal committed:
Remove new handler arguments from `wc_register_default_log_handler`

sirreal committed:
Fix settings log texts

Add 'woocommerce' to __ translated texts.
Improve texts.

sirreal committed:
Use 'notice' as level for deprecated WC_Logger::add messages.

Notice is the default (and recommended) threshold for production sites. Using
info, which is below notice, would result in ::add messages to be filtered by
default. This would be a confusing experience for users calling the deprecated
function and expecting a log.

sirreal committed:
Improve deprecation messages for WC_Logger

clear and remove *do* have alternatives. They have essentially
been moved to the file handler.

sirreal committed:
Improve doc comments and access.

sirreal committed:
Document _legacy context in WC_Log_Handler_File

sirreal committed:
Fix WC_Logger::add test to match new 'notice' level.

sirreal committed:
Deprecate wc_get_log_file_path

wc_get_log_file_path moves into `WC_Log_Handler_File` class as a static
method.
Update deprecations and version info.

sirreal committed:
Delay file log write when written too early

The function to look up log file paths depends on `wp_hash`, which is pluggable.
Log files cannot be found before `plugins_loaded` action. In case a log is
written to a file before `wp_hash` is available, cache and write on
`plugins_loaded` action.

sirreal committed:
Remove unneeded setUp and tearDown

WC_Logger tests no longer depend on writing/reading from log files.

sirreal committed:
Update use of deprecated wc_get_log_file_path

mikejolley commented on includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -0,0 +1,246 @@
+				case 'timestamp':
+				case 'tag':
+				case 'level':
+					$order_by = $_REQUEST['orderby'];
```
Sanitize this

mikejolley commented on includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -0,0 +1,246 @@
+			"SELECT log_id, timestamp, level, message, tag
+				FROM {$wpdb->prefix}woocommerce_log
+				WHERE 1 = 1 {$level_filter} {$search}
+				ORDER BY {$order_by} {$order_order} " .
```
These should be wrapped in esc_sql for safety.

sirreal replied:
I thought the whole method could be cleaner so I refactored. Have another look please.

mikejolley commented on includes/admin/views/html-admin-page-status-logs-db.php:
```diff
@@ -0,0 +1,30 @@
+<form method="get" id="mainform" action="">
+
+	<?php $log_table_list->views(); ?>
+	<?php $log_table_list->search_box( __( 'Search tag', 'woocommerce' ), 'tag' ); ?>
```
Is `tag` the correct terminology here? Confused me a moment due to WP tags taxonomy.

sirreal replied:
Yes, `tag` is probably best avoided. `source` might be a good solution, it indicates that you should tell where your log comes from. Other generic taxonomy names might be `group`, `package`, `bundle`...

mikejolley replied:
`source` sounds better to me.

mikejolley commented on includes/admin/views/html-admin-page-status-logs-db.php:
```diff
@@ -0,0 +1,30 @@
+	<input type="hidden" name="page" value="wc-status" />
+	<input type="hidden" name="tab" value="logs" />
+
+	<?php submit_button( __( 'Flush all logs', 'woocommerce' ), 'delete', 'flush-logs' ); ?>
```
Flush or 'clear'?

mikejolley commented on includes/class-wc-logger.php:
```diff
@@ -1,161 +1,230 @@
+	 * @param string $message
+	 * @return bool
+	 */
+	public function add( $handle, $message ) {
```
Since we have 300 extensions, I wonder if we can take a more laid back approach to this method. i.e. assume this is equivalent to debug level notice. Then extensions can log debug messages as is without using the new log features if they don't need to?

mikejolley replied:
As in, don't deprecate it :)

sirreal replied:
Ok, what about mentioning deprecation in the doc comment:

> This is not the preferred method for adding log messages. Please use log() or any one of the level methods (debug, info, etc.). This method may be deprecated in the future.

mikejolley commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,185 @@
+ * @category       Class
+ * @author         WooThemes
+ */
+class WC_Log_Handler_Email extends WC_Log_Handler {
```
This one was not in the settings. On purpose? How does one enable this?

sirreal replied:
It wasn't included in settings because it's not a traditional way of handling logs. It's not something that should be used without understanding. Email logs won't be colocated or persisted. You probably don't want to get emails for all of your log messages, so this handler has its own internal threshold. It should really be used in addition to one of the other handlers which are more traditional loggers and persist logs in 1 location.

As far as activation, either via the filter `woocommerce_register_log_handlers`, or by instantiating your own logger:

```php
// add to default handler stack:
add_filter( 'woocommerce_register_log_handlers', function( $handlers ) {
	array_push( $handlers, new WC_Log_Handler_Email() );
	return $handlers;
});
$logger = wc_get_logger();
// OR Logger with only your declared handlers
$logger = new WC_Logger( array( new WC_Log_Handler_Email() ) );

// Then log stuff!
$logger->emergency( 'Send me an emergency email!!!' );
```

mikejolley commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,185 @@
+	 */
+	protected static function get_subject() {
+		$site_name = get_bloginfo( 'name' );
+		return sprintf( __( '[%s] WooCommerce log messages', 'woocommerce' ), $site_name );
```
How about including some more information here to allow filtering, such as the number of logs or the highest severity?

mikejolley commented:
Nice set of tests :) Some failed but I think thats down to master. When updated from master it should be fine.

General questions/feedback below.

1. Having the log level and handler in general settings so prominent seems like a bad fit. I don't think store owners will understand this setting, or need it. Could they be moved elsewhere or set via constants in wp-config.php? Or do we need a new settings section?

I wouldn't mind constants with documentation in our WIKI.

2. It was confusing that I would enable debugging for say a gateway, and then find that no logs had been added due to the logging level setting.

In the end I found this was due to the level setting.

For things explicitly logged by extensions (which usually have their own settings) it makes less sense to have a level. 

Should we log `info` by default? 

-- OR -- 

Should we look at removing settings from gateways such as `Enable logging`? 

Feedback needed here I think. This through me off :)

3. The logger table could use human readable dates, a wider message column, and maybe icon based severity (not important, just a small improvement).

Also, the tag search - would this be better as dropdown of all available tags to offer filtering?

sirreal commented:
@mikejolley Thanks for the feedback 😀 

## 1. Settings location

I agree, general is not a good place for this setting. It doesn't seem to fit with any of the other existing tabs either. Are there other settings that would make sense on a `Developer` settings tab, now or in the future?

I think there is an advantage to making these settings available here rather than via constants, for example the first step in support could be "configure these log settings (file/debug) so we have a record of what's happening." That's more difficult to ask a user to do via code.

## 2. Default threshold

Picking defaults is hard. Should everything be logged by default, should the defaults be safe, or should the default be geared toward a production site? I chose `notice` as the default thinking about production sites (`notice` is the lowest level that represents a significant condition). When I updated calls to the logger, most of them seemed to be `info` level, which are ignored by the default I picked. Clearly, that is confusing if you're used to and expecting to start seeing logs immediately. Some third party perspective would be helpful to provide a coherent default experience for users.

Aside: Originally, I set `add` to log with `info` level, and realized that would be confusing when `add` messages were ignored by default, so changed `add` to log with `notice` level, processed by the default threshold. 

I believe gateways and extensions should be discouraged from disabling logging at this point. If the log threshold setting is easy to find and change, that should determine whether we log messages globally. I *always* want to know about an `emergency` level log. What other kinds of log related settings to extensions have?

## 3. Logger table

Great ideas about the columns. Regarding tag search, any idea how many log files exist on a typical site? Is there danger of a <select> being really long, or is there advantage to fuzzy search, e.g. matching `paypal-standard` and `paypal-something_else`?

mikejolley commented:
1. 

> I think there is an advantage to making these settings available here rather than via constants, for example the first step in support could be "configure these log settings (file/debug) so we have a record of what's happening." That's more difficult to ask a user to do via code.

That is true, however we do have a default logger so I don't anticipate this being a support issues. We've already established this is for developers, and developers don't need a settings UI?

2. You mention production sites - when it comes to logging I don't think we should limit whats logged by default. We're not talking about PHP error logs/notices/warnings here, we're talking about private logs the admin can inspect. If there were notices visible to users I'd agree with you,  but thats not the loggers purpose?

3. I don't think there would be too many - I mean, logs are usually created by plugins, so thats 1 log file per plugin potentially. I doubt that reaches high numbers?

sirreal commented:
@mikejolley Points 1 and 2 together make perfect sense. To clarify, I think the motivation for filtering logs is really about volume and noise, but if that hasn't been a problem before now, it's unlikely to become a problem in the near future. Issues related to volume should also be improved by log rotation.

sirreal committed:
Replace log `tag` with log `source`

sirreal committed:
Do not deprecate WC_Logger::add

sirreal committed:
Improve email handler subject

sirreal committed:
Format date in log viewer table

sirreal committed:
Fix style

sirreal committed:
Rework db log table

Change tag from text search to select list.
Flatten and clean up query logic.
Add required methods for source <select>.

sirreal committed:
Remove logging

sirreal committed:
Use constants over settings to configure logger

Settings have been removed.
WC_LOG_THRESHOLD and WC_LOG_HANDLER constants have been added.

sirreal committed:
Fix tests for different threshold configuration

No filtering by default.
Options configuration has been removed.

sirreal committed:
Use plural in multiple log email

shivapoudel commented on includes/admin/class-wc-admin-status.php:
```diff
@@ -77,7 +77,7 @@ public static function get_tools() {
 	 */
 	public static function status_logs() {
-		if ( 'db' === get_option( 'woocommerce_default_log_handler' ) ) {
+		if ( defined( 'WC_LOG_HANDLER' ) && 'WC_Log_Handler_DB' === WC_LOG_HANDLER ) {
```
👍

sirreal committed:
Define column sizes for log table

sirreal commented:
Log configuration is removed from settings and available by constants (or via constructors for more control). `WC_LOG_HANDLER` expects the class name of the handler you want to use. `WC_LOG_THRESHOLD` expects a level. By default, everything is logged by file.

```php
define( 'WC_LOG_HANDLER', 'WC_Log_Handler_DB' );
define( 'WC_LOG_THRESHOLD', 'emergency' );
```

I've made changes to deal with the rest of the issues we've discussed, except:

> maybe icon based severity (not important, just a small improvement)

8 levels, which are relatively abstract concepts, is near impossible to represent well as icons. We could use fewer icons each describing the severity of a group of levels, but I'm not sure that adds much relevant information.

mikejolley commented:
> 8 levels, which are relatively abstract concepts, is near impossible to represent well as icons. We could use fewer icons each describing the severity of a group of levels, but I'm not sure that adds much relevant information.

Ok agreed. Lets just shrink that column down then, and perhaps even color code so it's easily scannable :)?

sirreal committed:
Colorize leveles in db log table

sirreal commented:
@mikejolley 2 alternatives. Rather than icons, we could use the first letter to represent level along with color coding. Color coded by severity:

* emergency, alert: red
* critical, error: orange
* warning, notice: yellow
* info: blue
* debug: green

Examples are the smallest width before the table collapses.

# By level letter
Full level in `title` attribute. This saves about 8% width.

![letter](https://cloud.githubusercontent.com/assets/841763/21481297/6c3c16f4-cb66-11e6-9367-259a957c4b34.png)

# By level word
![word](https://cloud.githubusercontent.com/assets/841763/21481298/6c7f5cde-cb66-11e6-84f4-470b09678032.png)

nb commented on includes/class-wc-logger.php:
```diff
@@ -64,16 +64,16 @@ public function should_handle( $level ) {
 	 */
-	public function add( $handle, $message ) {
-		wc_deprecated_function( 'WC_Logger::add', '2.7', 'WC_Logger::log' );
+	public function add( $handle, $message, $level=WC_Log_Levels::NOTICE ) {
```
Tiny, but I think we need spaces around the equal sign.

sirreal replied:
I think that's from an outdated commit, the current branch looks correct: https://github.com/woocommerce/woocommerce/blob/wc-logger/includes/class-wc-logger.php#L81

sirreal commented:
Closed in favor of #12716.
